### ISSUE TITLE
LoopHelper FPS limiter lags if work is done outside of loop

### ISSUE THREAD

**Issue by nyanpasu64** [alexheretic/spin-sleep#5]
`spin_sleep::LoopHelper` is designed wrong. It doesn't use the end `loop_sleep()` of the previous frame to time the end of the next frame... but uses the beginning `loop_start()` of the next frame (which is a different Instant).

I have a app using Winit to create a GUI. It renders frames in `Event::MainEventsCleared`, using wgpu in PresentMode::Immediate (no vsync). I'm trying to throttle the frame rendering rate using LoopHelper.

When I set the target FPS to 250, I get 220 in practice. When I set the target FPS to 200, I get 180 in practice. This is because winit does work between callback/render iterations.

If I move `loop_sleep()` to right before `loop_start()` (contrary to what the example code does), I get 198 FPS, which is still systematically lower than 200.

The code causing this is this variable and its usages:

https://github.com/alexheretic/spin-sleep/blob/929eef77f8282998cc33addb26de406d54c53427/src/loop_helper.rs#L46

Not sure how to program it. If the rendering is faster than the target FPS (the limiter is sleeping), each `loop_sleep()`'s "target endpoint" should be exactly `target_delta` after the previous frame's "target endpoint", rather than the `Instant` returned after the previous frame's sleep ends or when the current frame begins. If the rendering is slower than the target FPS (each frame takes longer than `target_delta`), then you need to use the `Instant` returned after the previous frame's sleep is called. But I don't know if you can distinguish these cases.

Basically I'm trying to recreate `setInterval()` but synchronously via thread blocking, rather than as a callback. I may fork spin-sleep and try to implement the API myself. Or I might use winit timers instead (https://github.com/rust-windowing/winit/blob/master/examples/timer.rs). Don't know which to pick, tbh.

**Comment by alexheretic** [comment]
I'm not sure it's a problem with the API & more that you're not calling `loop_start()` at the start of an actual loop. There should be nothing between `loop_start()` and `loop_sleep()` they're separate calls to allow delta to be returned at the start of a loop and the sleeping to be done at the end.

Winit's `EventLoop:run` isn't a loop so can't be used the same way. If you call `loop_start()` at the start of `MainEventsCleared` it's like calling it in the middle of the traditional loop. However, if you don't need the delta at the start it can still work fine by just calling it all at the end. Here's how I was doing it in a wgpu project

```rust
event_loop.run(move |event, _, control_flow| {
    use winit::{event::*, event_loop::ControlFlow};
    *control_flow = ControlFlow::Poll;

    match event {
        // handle other events

        Event::MainEventsCleared => {
            // compute stuff

            // render stuff

            if let Some(fps) = limiter.report_rate() {
                // fps
            }
            limiter.loop_sleep(); // or `loop_sleep_no_spin()` to save battery
            limiter.loop_start();
        }
    }
})
```
This works as **all** work is being done correctly between _start_ & _sleep_. 

Tbf once we don't need the delta at the start of the loop the API makes less sense & could be streamlined. But it still works here & `LoopHelper` _isn't_ `WinitEventLoopHelper`, so I think the API is working as designed.

If having `loop_start()` at the end bothers you, you could just write a specialised helper or wrapper to improve it.

**Comment by nyanpasu64** [comment]
> If I move `loop_sleep()` to right before `loop_start()` (contrary to what the example code does), I get 198 FPS, which is still systematically lower than 200.

I still get this discrepancy (with my project in debug mode and dependencies in release mode) regardless if I place the two adjacent function calls before or after `// compute/render stuff`.

I kinda feel we're not on the same page regarding the design of the API. Should I fork the repo and build it towards my own design?

**Comment by alexheretic** [comment]
For my usage the example works accurately, debug or release. Can you post a
non-working example?

In any case perhaps it's best for you to just build your own loop helper
abstraction and possibly just use the sleep abstraction from this crate.

The loop helper is not a lot of code so if it's not working for you just
roll your own 🙂

I'd rather avoid breaking changes to this crate if I can. If you like you
could propose your design as a `LoopHelper2` (or similar new thing) pr and
we could take it from there.

**Comment by nyanpasu64** [comment]
I'll make such a PR at some point, I'm currently writing some atomic/unsafe code unrelated to frame pacing, so my project doesn't compile.

The project where I'm having trouble is https://github.com/nyanpasu64/spectro2. Specifically commit https://github.com/nyanpasu64/spectro2/commit/f430b7d38ea74b1f13a9be181cd9bf270936e095 hits 198 FPS, and its parent https://github.com/nyanpasu64/spectro2/commit/268bb0d717667a7cf523a1a6318760afea6d93a1 (which does winit work between loop iterations) hits 180 FPS.

**Comment by alexheretic** [comment]
I eventually came to the conclusion that i agree that the loop helper api is flawed. But i also think it doesn't belong in this crate at all and to instead focus the crate on spin sleeping functionality only.

So it is kinda "very softly" deprecated by removing it from the readme & docs index. I'd like to fully deprecate it if there is a drop in replacement crate available elsewhere.

**Comment by alexheretic** [comment]
#19
